Execution Environments

shanemcd opened the pull request:
This is a long-running feature branch where Execution Environments is being implemented. A proper upgrade plan for the community will need to be in place before this can be merged and included in a release of AWX.


## TODOs
- [x] Documentation
- [x] EE Module in AWX collection
- [x] Integrate Receptor

jbradberry committed:
Add a new ExecutionEnvironment model

jbradberry committed:
Create the RBAC access class for execution environments

jbradberry committed:
Add in the basic list and detail api views

jbradberry committed:
Change OrganizationSerializer to show and set default_environment

jbradberry committed:
Add execution_environment and pull to the fields for UJs and UJTs

jbradberry committed:
Expose an API view for all of the execution environments under an org

jbradberry committed:
Expose execution environments in awxkit and awx-cli

jbradberry committed:
Add a new Swagger topic

jbradberry committed:
Make changes to support capture by the activity stream

Including exposing a new API view for a particular EE's activity
stream objects.

jbradberry committed:
Add a sublist api view for the UJTs that use a given execution environment

shanemcd committed:
Initial EE integration

shanemcd committed:
Initial EE integration

jbradberry committed:
Run a receptor node in the dev environment

nixocio committed:
Add execution environments files (#7909)

Update navigation bar and routing system to add execution environments.
Also, add stub files for the remaining related work.

See: https://github.com/ansible/awx/issues/7885
Also: https://github.com/ansible/awx/issues/7884

AlanCoding committed:
Add Z to volume mount

Set ansible-runner back to main fork due to merge

AlanCoding committed:
Add Z to volume mount

Update to AWX execution environment
  use the special 2.9 container image
  revert setting back for merge

Fix another permission error by mapping 2 folders
  also create folders before running

AlanCoding committed:
Purge environment variables to work with ansible-runner changes

Remove inventory scripts show because they no longer exist

Remove reference to non-existent callback directory

Remove more references to removed paths

AlanCoding committed:
Respect user proot show paths when using containers

AlanCoding committed:
Fix pathing issue with custom credentials

also fix some minor flake8 issues

AlanCoding committed:
Fix pathing issue for credential file references

AlanCoding committed:
Fix project folder deletion

Fix another absolute path reference in containers

AlanCoding committed:
Deal with missing HOME env var

AlanCoding committed:
Initial minimal hooking up of JT EEs to jobs

nixocio committed:
Add list Execution Environments (#8148)

See: https://github.com/ansible/awx/issues/7886

AlanCoding committed:
Changes to get execution environments factories working (#8126)

nixocio committed:
Add feature to Add/Edit Execution Environments (#8165)

* Add feature to Add/Edit Execution Environments

Add feature to Add/Edit Execution Environments.

Also, add key for `ExecutionEnvironmentsList`.

See: https://github.com/ansible/awx/issues/7887

* Update registry credential label

nixocio committed:
Add details page for Execution Environments (#8172)

* Add feature to Add/Edit Execution Environments

Add feature to Add/Edit Execution Environments.

Also, add key for `ExecutionEnvironmentsList`.

See: https://github.com/ansible/awx/issues/7887

* Add details page for execution environments

Add details page for execution environments

See: https://github.com/ansible/awx/issues/8171

jbradberry committed:
Expose the user capabilities dict for EEs (#8208)

nixocio committed:
Update usage of summary_fields for execution environments (#8217)

Update usage of summary_fields for execution environments. Also, update
unit-tests to cover this change.

See: https://github.com/ansible/awx/issues/8216

AlanCoding committed:
Collections volume permission fix, and container group fix

Use same image for both types of container isolation

Inventory move fix related to container groups

nixocio committed:
Add organization as part of creating/editing an execution environments

Add organization as part of creating/editing an execution environments

If one is a `system admin` the Organization is an optional field. Not
providing an Organization makes the execution environment globally
available.

If one is a `org admin` the Organization is a required field.

See: https://github.com/ansible/awx/issues/7887

nixocio committed:
Add execution environment list to Organizations

Add execution environment list to Organizations

See: https://github.com/ansible/awx/issues/8210

shanemcd committed:
REVERT ME: Install ansible/devel for now

shanemcd committed:
Use WIP version of collections_requirements.yml

Pulled from ansible-builder/test/data/awx

shanemcd committed:
Install receptorctl in awx venv

shanemcd committed:
Add new ee container

shanemcd committed:
Add AWX EE definition

shanemcd committed:
Use official receptor image

shanemcd committed:
REVERT ME: Install community.general in image

This is needed for the wait_fors in the launch scripts to work

shanemcd committed:
Add license for receptor

shanemcd committed:
Back to green

shanemcd committed:
Remove "pull" field from EE mixin

I think this should go on the EE definition itself

shanemcd committed:
WIP: Module for EEs

shanemcd committed:
Fix linter

shanemcd commented:
recheck

jakemcdermott commented:
recheck

jakemcdermott commented:
regate

jakemcdermott commented:
recheck

shanemcd committed:
Begin integrating receptor

shanemcd committed:
Set SDB_NOTIFY_HOST for all processes

jbradberry committed:
Add in more model changes around execution environments

- a new unique name field to EE
- a new configure-Tower-in-Tower setting DEFAULT_EXECUTION_ENVIRONMENT
- an Org-level execution_environment_admin_role
- a default_environment field on Project
- a new Container Registry credential type
- order EEs by reverse of the created timestamp
- a method to resolve which EE to use on jobs

shanemcd committed:
Create receptor mesh in cluster development environment

shanemcd committed:
Actually cancel things

jbradberry committed:
Fill in the new execution environment collection module

as well as changes to other ones that need to be able to attach EEs.

jbradberry committed:
Make use of the EE resolver code when launching jobs

jbradberry committed:
Move the resolve_execution_environment method to the mixin class

so that it can be used with AdHocCommands as well.

shanemcd committed:
Surface errors when launching jobs through Receptor

This will raise errors such as:

exec: "ansible-runner": executable file not found in $PATH

shanemcd committed:
Pull awx -> receptor job code into its own class

shanemcd committed:
FOLLOW UP ON THIS: Fix fact_cache directory location

The part where we pass in the runner params to the processor phase is
legit. Need to investigate why the fact_cache directory is no longer nested
under job.id.

shanemcd committed:
Fix tests after rebasing in inventory update refactor

shanemcd committed:
WIP: containerized inventory updates. Thanks ALAN!!

AlanCoding committed:
More inventory update containerization fixes

AlanCoding committed:
Make insights integration tests pass again

jbradberry committed:
Populate the EE name field in awxkit

jbradberry committed:
Adjust ExecutionEnvironmentAccess to account for the new EE admin role

AlanCoding committed:
Add utility method to get controller private_data_dir

AlanCoding committed:
Add back in the subversion requirement

jbradberry committed:
Ensure that a fallback EE is available to be found

for the failing tests.

AlanCoding committed:
Change the shebang back to just python

jbradberry committed:
Ensure that the updated job instance is used

when attaching an EE.

AlanCoding committed:
Remove files moved to the ansible/awx-ee repo

These have been moved to:

https://github.com/ansible/awx-ee

that will be the home for the processes needed to
build this execution environment.

jbradberry committed:
Insert a default EE into the development environment

jbradberry committed:
Make sure that the new credential type is in the choices list

shanemcd committed:
Show EE images that are managed by tower in UI

jbradberry committed:
Improve the behavior of EE resolution for ad hoc commands

- call resolve_execution_environment during AdHocCommand.save()
- wrap the fallback call of the resolver in tasks.py in disable_activity_stream()

AlanCoding committed:
Change the default EE location

AlanCoding committed:
Fix collection pep8 failure

AlanCoding committed:
Fix another credential path-in-container bug

AlanCoding committed:
Fix another svn issue due to pre-existing folder

AlanCoding committed:
Fix yet another host vs container path bug

shanemcd commented:
@AlanCoding looks like you may have broken the build with that last commit. Can you take a look? Maybe we should go back to only doing PRs to this branch.

AlanCoding committed:
Fix unit test fallout

AlanCoding committed:
Fix raw archive project updates

Several squashed commits

Fix git bug introduced by setting remote tmp in project path
change shebang back to py3 again
Revert shebang change

AlanCoding committed:
Allow jobs to run in the base ansible-runner image (#8949)

shanemcd committed:
Get podman-in-docker working under cgroups v2

shanemcd committed:
Update receptor config to allow for runtime options

shanemcd committed:
InstanceGroup#is_containerized -> InstanceGroup#is_container_group

shanemcd committed:
UnifiedJob#is_containerized -> UnifiedJob#is_container_group_task

shanemcd committed:
Deleted unused build_params_process_isolation method

shanemcd committed:
Delete some old container group v1 code

shanemcd committed:
Use official awx-ee by default

shanemcd committed:
Minimally functional container group v2 w/ receptor

shanemcd committed:
flake8

shanemcd committed:
Make receptorctl easier to use in dev env

shanemcd committed:
Dont zip/unzip private data dir for local jobs

shanemcd committed:
Create awx group in container

shanemcd committed:
Add PATH to blocked inventory source vars

This used to be skiped because PATH was already present in the env we
constructed for runner.

AlanCoding committed:
Use the fully qualified inventory plugin name only for foreman

AlanCoding committed:
Fix minor syntax error failing AdHocCommands

shanemcd commented:
recheck

AlanCoding committed:
Fix obvious code error with foreman inventory

ryanpetrello committed:
introduce a new global flag for denoating K8S-based deployments

- In K8S-based installs, only container groups are intended to be used
  for playbook execution (JTs, adhoc, inventory updates), so in this
  scenario, other job types have a task impact of zero.
- In K8S-based installs, traditional instances have *zero* capacity
  (because they're only members of the control plane where services
  - http/s, local control plane execution - run)
- This commit also includes some changes that allow for the task manager
  to launch tasks with task_impact=0 on instances that have capacity=0
  (previously, an instance with zero capacity would never be selected
  as the "execution node"

This means that when IS_K8S=True, any Job Template associated with an
Instance Group will never actually go from pending -> running (because
there's no capacity - all playbooks must run through Container Groups).
For an improved ux, our intention is to introduce logic into the
operator install process such that the *default* group that's created at
install time is a *Container Group* that's configured to point at the
K8S cluster where awx itself is deployed.

AlanCoding committed:
Exception handling to always release work units

shanemcd committed:
Update ExecutionEnvironments.jsx with breadcrumb replacement

shanemcd commented:
recheck

shanemcd commented:
recheck

shanemcd committed:
Pin to latest version of PyYAML

Fixes https://github.com/yaml/pyyaml/issues/478

yagomarques committed:
WIP Inclusion of the EE option in the payloads within the Organization and Projects. (#9145)

* add ee option on factories for organizations

* add new lines

* remove inventory from the options

* remove line

* remove line from projects

* fix the tuple

* fix lint problems

AlanCoding committed:
Make sure project updates run in default EE (#9172)

* Make sure project updates run in default EE

* Remove project execution_environment field from collection

AlanCoding committed:
Assure that unit_id is always defined (#9180)

rebeccahhh committed:
execution model pull container options added

rebeccahhh committed:
migration for container options for EE model

Co-authored-by: Shane McDonald <me@shanemcd.com>

shanemcd committed:
Provide new default pod defintion in CG metadata (#9181)

shanemcd committed:
Set correct SDB_NOTIFY_HOST in minikube env

shanemcd committed:
Stop and kill dispatcher and callback reciever as group

AlanCoding commented on awx/api/metadata.py:
```diff
@@ -209,7 +209,7 @@ def determine_actions(self, request, view):
 
                 if field == "pod_spec_override":
-                    meta['default'] = PodManager().pod_definition
+                    meta['default'] = AWXReceptorJob().pod_definition
```
It would be nice if this pod definition was not defined in 2 places to be DRY.

AlanCoding commented on awx/main/access.py:
```diff
@@ -1308,6 +1308,53 @@ def can_unattach(self, obj, sub_obj, relationship, *args, **kwargs):
+    def can_add(self, data):
+        if not data:  # So the browseable API will work
+            return Organization.accessible_objects(self.user, 'execution_environment_admin_role').exists()
+        return self.check_related('organization', Organization, data)
```
We should test that if the organization is null, a non-superuser can not add the execution environment. It might be, I just wanted to make the obvious note. I don't think `check_related` does this, that's what the `mandatory` kwarg was for.

AlanCoding commented on awx/main/access.py:
```diff
@@ -1308,6 +1308,53 @@ def can_unattach(self, obj, sub_obj, relationship, *args, **kwargs):
+        if self.user not in obj.organization.execution_environment_admin_role:
+            raise PermissionDenied
+        org_pk = get_pk_from_dict(data, 'organization')
+        if obj and obj.organization_id != org_pk:
```
covering these kinds of cases was why we have the `check_related` method to begin with

AlanCoding commented on awx/main/access.py:
```diff
@@ -1308,6 +1308,53 @@ def can_unattach(self, obj, sub_obj, relationship, *args, **kwargs):
+        if obj and obj.organization_id is None:
+            raise PermissionDenied
+        if self.user not in obj.organization.execution_environment_admin_role:
+            raise PermissionDenied
```
Other methods just return `True` in these cases. I'm not against changing them all to raising instead, but it's inconsistency for no reason.

AlanCoding commented on awx/main/migrations/0125_more_ee_modeling_changes.py:
```diff
@@ -0,0 +1,46 @@
+# Generated by Django 2.2.16 on 2020-11-19 16:20
```
are we going to combine these migrations into 1 file before merge?

nixocio committed:
Mark string to translation

Mark string to translation

shanemcd committed:
Get system jobs working under new deployment model (#9221)

shanemcd committed:
Create default EE as a part of create_preload_data

shanemcd committed:
Fix missing postgresql module

jakemcdermott committed:
Add an interface for new ee options

jakemcdermott committed:
Remove client-side url validator

shanemcd committed:
Quick fix for jobs failing in dev environment / VMs

The other alternative here is to go all the way with
https://github.com/ansible/ansible-runner/pull/617, which is proving to be
difficult if not impossible.

shanemcd committed:
Update default ee image

shanemcd committed:
Update receptor.conf in dev env

shanemcd committed:
Remove Ansible from control plane

Execution Environments or bust!

shanemcd committed:
Install Ansible only for collection tests

nixocio committed:
Migrate EE list to tables

Migrate EE list to tables.

See:https://github.com/ansible/awx/issues/7884

nixocio committed:
Update Job Detail container group variable

`is_containerized` was updated on the API side to be
`is_container_group`.

shanemcd committed:
Handle quota exceeded in Container Groups v2

rebeccahhh committed:
changed the field name from 'container_options' to simply 'pull'

nixocio committed:
Add Execution Environments into a few screens

Add EE to the following screens:

* Job Template
* Organization
* Project
* Workflow Job Template

Also, add a new lookup component - ExecutionEnvironmentLoookup.

See: https://github.com/ansible/awx/issues/9189

shanemcd committed:
Container Pull Option -> Pull

Co-authored-by: Jake McDermott <yo@jakemcdermott.me>

rebeccahhh committed:
adding needed url endpoint for copy functionality and the beginning of some testing that can be fleshed out more fully in later work

rebeccahhh committed:
add a functional test for creating an EE, remove bum copy function because it's not needed, copy works from the base class

moved AWXKit pull additions to separate PR and made some changes that were causing linting errors in tests and add copy to show_capabilities for the ee serializer

unlikelyzero commented:
@nixocio based on some test PR feedback, would it be possible to replace the data-cy attribute with a data-ouia attribute?

shanemcd committed:
Add migration to reset custom pod specs

rebeccahhh committed:
default pull options for container images to None, also adding pull options to awxkit

beeankha committed:
Enable utilized EE Collections name and version info to be detected

rebeccahhh committed:
edit original migration file, add blank string as acceptable to model

rebeccahhh committed:
add org admins as able to control EEs even if they don't have the ee_admin role for the specific ee and prevent managed_by_tower EEs from being edited/deleted

nixocio committed:
Update selectors on EE details page to ease testing

Update selectors on EE details page to ease testing.

beeankha committed:
Assign entire file path to variable

rebeccahhh committed:
add execution_environment_admin_role to the an organizations read role, which access.py uses for determining access to reading an ee within an organization,

add migration file for execution_env_admin role addition to read_roles within an organization,

and set check related to mandatory

jbradberry committed:
Make the managed_by_tower field read-only for EEs (similar to how we deal with it not being settable for Credentials) and add permissions checking for Org EE Admins.

can_add: gets an explicit role to check against, `'execution_environment_admin_role'`
can_change: leverages `self.check_related()` for the case where the Org is not changing, but also adds an explicit check for the EE Admin Role when the Org is changing to an explicit different Org.

kdelee committed:
make sure we use built in credential type

this way we can pass kind="registry" to akit creat method and
we get the correct built in type

shanemcd committed:
Use official Receptor 0.9.5 release

beeankha committed:
Add installed_collections column to unified job query

jakemcdermott committed:
Add name field

jbradberry committed:
Turn off permissions check bypassing for admins when hitting the execution environment list and detail views.

beeankha committed:
Enable EE collections info to be loaded as valid JSON vs stringified JSON

beeankha committed:
Bump up unified_jobs_table version for new column addition

shanemcd committed:
Fix receptor.conf path in dev env

shanemcd committed:
Wire up --pull option for EEs

shanemcd committed:
Add "copy" to EE related links

nixocio committed:
Minor update EE tables

* Add table header `actions`
* Add `name` as default search

See: https://github.com/ansible/awx/issues/7884
Also: https://github.com/ansible/awx/issues/9087

nixocio committed:
Add organization to EE details page

Add organization to EE details page.

See: https://github.com/ansible/awx/issues/9432

nixocio committed:
Update EE breadcrumb to use name instead of image

Update EE breadcrumb to use name instead of image to be consistent with
the other screens.

See: https://github.com/ansible/awx/issues/9087

shanemcd committed:
Fix mode for k8s launch scripts

shanemcd committed:
Only install podman in local dev env

nixocio committed:
Add EE to inventory sources

Add EE to inventory sources

See: https://github.com/ansible/awx/issues/9189

nixocio committed:
Linkify reference to EE on details page

Linkify reference to EE on a few details page.

See: https://github.com/ansible/awx/issues/9189

shanemcd committed:
Add ui_next to .dockerignore

shanemcd committed:
Use var instead of set_fact

shanemcd committed:
Use receptor 0.9.6

shanemcd committed:
Dont require credential for Container Groups

Might want to follow up and make this only apply to Tower on Kubernetes

shanemcd committed:
For container group pods, use namespace Tower is deployed into by default

shanemcd commented:
recheck
